### DISCUSSION TITLE
Proper usage of combined ImageSamplers with Vulkan

### DISCUSSION THREAD

**Discussion by wpsimon09** [shader-slang/slang#5654] [Q&A] [answered]
Hello, 

I have recently rewrote all my shaders from `GLSL` to Slang. I am using Vukan as a primary API . To pass texture to the Fragment shader I am using push descriptors that are of type `vk::DescriptorType::eCombinedImageSampler`. 

To accommodate for this i have created the following struct in my Slang shader 

```c
struct TextureSampler
{
    Texture2D texture;
    SamplerState sampler;

    float4 Sample(float2 uv) {
        return texture.Sample(sampler, uv);
    }
}

[[vk::binding(2,0)]]
ConstantBuffer<TextureSampler> albedo;
```

This works as expected. However Vulkan validation layers are complaining with the following message.

```txt
Error: { Validation }:
	messageIDName   = <VUID-VkGraphicsPipelineCreateInfo-layout-07988>
	messageIdNumber = 559874765
	message         = <Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-layout-07988 ] Object 0: handle = 0x340000000034, type = VK_OBJECT_TYPE_SHADER_MODULE; Object 1: handle = 0x350000000035, type = VK_OBJECT_TYPE_PIPELINE_LAYOUT; | MessageID = 0x215f02cd | vkCreateGraphicsPipelines(): pCreateInfos[1].pStages[1] SPIR-V (VK_SHADER_STAGE_FRAGMENT_BIT) uses descriptor [Set 0, Binding 3, variable "albedo.sampler"] (type VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER or VK_DESCRIPTOR_TYPE_SAMPLER) but was not declared in the pipeline layout.
The Vulkan spec states: If a resource variables is declared in a shader, a descriptor slot in layout must match the shader stage (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-layout-07988)>
	Objects:
		Object 0
			objectType   = ShaderModule
			objectHandle = 57174604644404
		Object 1
			objectType   = PipelineLayout
			objectHandle = 58274116272181

``` 

My question is if there is proper way of using `vk::DescriptorType::eCombinedImageSampler` without my walk around solution ? And also how come that this works ? 

What I was thinking is that I will pass the sampler and Texture image separately and than use the same sampler for all of the images but I would prefer to use the combined image sampler if possible.

Thank you :)

**Comment by jkwak-work** [comment, marked as answer]
The provided shader code looks fine to me. I think the validation error is coming from the other part of the shader where the sampler is used in an unexpected stage. The struct might be meant for vertex shader only but it might be used in a fragment shader incorrectly.

If you think it is a bug on Slang, we can work on it, but we need a repro case that reproduces the validation error mentioned.

The texture-combined samplers such as `sampler2D` in GLSL are legalized by Slang in a way described above. If you use `sampler2D`, Slang will create a Texture2D and SamplerState automatically. Please give a try and let us know if it doesn't work.

*EDIT: I forgot to mention that you have to use `-allow-glsl` option when using GLSL types like `sampler2D`.

**Comment by wpsimon09** [comment]
Hello, thank you for taking time and replying to my question. 

I have used `sampler2D` type for my `ConstantBuffer<>` instead of my custom struct and it all worked exactly as expected and without any message(s) from validation layers.

Just to clarify, the `TextureSampler` struct that you have seen was used only in one fragment shader and since I am very new to the Slang (loving it so far, very good job) I have disabled all my other pipelines that use different shaders and/or Pipeline layouts in order to experiment a bit. The thing is that even with the validation layer saying the sampler is not being filled in, my custom solution worked and I **could see the texture being sampled** without an issue. I was just interested in why is that so.

**Reply by jkwak-work** [reply]
I am not sure if I understood the question correctly. The question is why did the sampling work even when the samplers were not properly filled. I am guessing that it is a kind of undefined behavior that may or may not work depending on how the graphics driver is implemented. But I wouldn't rely on any undefined behavior because it may work on one type of GPUs and may not work on other GPUs.

**Reply by wpsimon09** [reply]
Yes, you understood the question correctly. Thank you for your answer !
